### ISSUE TITLE
`\tcb@gettitle` should use `\protected@edef`

### ISSUE THREAD

**Issue by eg9** [T-F-S/tcolorbox#14]
If in the following example the redefinition of `\tcb@gettitle` is commented out, the LaTeX run breaks badly.

The fix is exactly using `\protected@edef` instead of `\edef`.

    \documentclass{article}
    \usepackage{amsfonts}
    \usepackage{tcolorbox}
    \tcbuselibrary{theorems}
    \newtcbtheorem{boxtheorem}{Theorem}{}{thm}
    
    \makeatletter
    \def\tcb@gettitle#1{\protected@edef\@currentlabelname{#1}}
    \makeatother
    
    
    \begin{document}
    
    \begin{boxtheorem}{This is the \emph{problem}: $\mathbb{N}$}{}
    This is fine: $\mathbb{N}$
    \end{boxtheorem}
    
    \end{document}

Reference: http://tex.stackexchange.com/questions/339636/clash-between-tcolorbox-and-amsfonts

**Comment by T-F-S** [comment]
Thank you for reporting this really annoying bug and also thank you for providing a solution.

I found a second problem which is related to this one. `\tcb@gettitle` is intended to become `\NR@gettitle`, if the `nameref` package is loaded. Unfortunately, this only happens, if `nameref` is loaded after `tcolorbox`.

The next version of `tcolorbox` will include your fix and also the fix for the related problem:

```latex
\def\tcb@gettitle#1{\protected@edef\@currentlabelname{#1}}

\AfterEndPreamble{%
  \ifdefined\NR@gettitle%
    \let\tcb@gettitle\NR@gettitle%
  \fi%
}

```

**Comment by eg9** [comment]
> On 17 Nov 2016, at 15:08, Thomas F. Sturm <notifications@github.com> wrote:
> 
> Thank you for reporting this really annoying bug and also thank you for providing a solution.
> 
> I found a second problem which is related to this one. \tcb@gettitle is intended to become \NR@gettitle, if the nameref package is loaded. Unfortunately, this only happens, if nameref is loaded after tcolorbox.
> 
> The next version of tcolorbox will include your fix and also the fix for the related problem:
> 
> \def\tcb@gettitle#1{\protected@edef\@currentlabelname
> {#1}}
> 
> 
> \AfterEndPreamble{%
> 
>   
> \ifdefined\NR@gettitle%
> 
>     
> \let\tcb@gettitle\NR@gettitle%
> 
>   
> \fi%
> 
> }

I quickly checked the other instances of \edef in tcolorbox.sty and I saw nothing
dangerous.

Thanks for all your work on this great package!

Ciao
Enrico

**Comment by T-F-S** [comment]
The fix is now incorporated into version `3.96 2016-11-18` which is on its way to CTAN.

**Comment by eg9** [comment]
Tested and appearing good!
